src/oskbd/linux.rs

```diff
@@ -47,121 +30,61 @@ impl KbdIn {
     }
 
     fn new_linux(dev_path: &Path) -> Result<Self, std::io::Error> {
-        let kbd_in_file = File::open(dev_path)?;
-        let mut kbd_in_dev = Device::new_from_fd(kbd_in_file)?;
+        let mut kbd_in_dev = Device::open(dev_path)?;
 
         // NOTE: This grab-ungrab-grab sequence magically
-        // fix an issue I had with my Lenovo Yoga trackpad not working.
-        // I honestly have no idea why this works haha.
-        kbd_in_dev.grab(GrabMode::Grab)?;
-        kbd_in_dev.grab(GrabMode::Ungrab)?;
-        kbd_in_dev.grab(GrabMode::Grab)?;
-
-        Ok(KbdIn { device: kbd_in_dev })
+        // fix an issue with a Lenovo Yoga trackpad not working.
+        // No idea why this works.
+        kbd_in_dev.grab()?;
+        kbd_in_dev.ungrab()?;
+        kbd_in_dev.grab()?;
+
+        Ok(KbdIn {
+            device: kbd_in_dev,
+            events: Default::default(),
+        })
     }
 
-    pub fn read(&self) -> Result<InputEvent, std::io::Error> {
-        let (status, event) = self
-            .device
-            .next_event(ReadFlag::NORMAL | ReadFlag::BLOCKING)?;
-        std::assert!(status == ReadStatus::Success);
-        Ok(event)
+    pub fn read(&mut self) -> Result<InputEvent, std::io::Error> {
+        while self.events.is_empty() {
+            self.device
+                .fetch_events()?
+                .into_iter()
+                .for_each(|ev| self.events.push_back(ev));
+        }
+        Ok(self.events.pop_front().expect("not empty"))
```

Comment:
```suggestion
        Ok(self.events.pop_front().expect("event queue is empty"))
```

Usually, this message tells us why it fails.